lib/charms/operator_libs_linux/v0/dnf.py

```diff
@@ -0,0 +1,462 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Abstractions for system's DNF package information and repositories."""
+
+import copy
+import itertools
+import os
+import shutil
+import subprocess
+from collections import deque
+from concurrent.futures import ThreadPoolExecutor
+from enum import Enum
+from functools import wraps
+from io import StringIO
+from typing import Callable, Dict, List, Literal, Optional, Union
+
+
+class _PackageState(Enum):
+    INSTALLED = "installed"
+    AVAILABLE = "available"
+    ABSENT = "absent"
+
+
+class DNFExecutionError(Exception):
+    """Raise when dnf encounters an execution error."""
+
+
+def _dnf(
+    command: str,
+    args: Optional[Union[Union[str, os.PathLike], List[Union[str, os.PathLike]]]] = None,
+    preargs: Optional[List[str]] = None,
+    postargs: Optional[List[str]] = None,
+) -> subprocess.CompletedProcess[str]:
+    """Execute DNF command.
+
+    Args:
+        command (str): Command to execute.
+        args (Optional[Union[Union[str, os.PathLike], List[Union[str, os.PathLike]]]]):
+            Arguments to operate on. (Default: None).
+        preargs (Optional[List[str]], optional):
+            Optional arguments to prepend before command. (Default: None).
+        postargs (Optional[List[str]], optional):
+            Optional arguments to append after command. (Default: None).
+
+    Raises:
+        DNFExecutionError: Raised if specified DNF command fails to execute.
+
+    Returns:
+        (subprocess.CompletedProcess[str]):
+            returncode, stdout, and stderr of completed DNF command.
+    """
+    args = [args] if type(args) == str else args if type(args) == list else []
+    preargs = preargs if preargs is not None else []
+    postargs = postargs if postargs is not None else []
+    cmd = ["dnf", "-y", *preargs, command, *postargs, *args]
+    try:
+        return subprocess.run(
+            cmd,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            universal_newlines=True,
+        )
+    except subprocess.CalledProcessError as e:
+        raise DNFExecutionError(
+            f"Failed to perform command {' '.join(cmd)}. Reason:\n\n{e.stderr}"
+        )
+
+
+class PackageInfo:
+    """Dataclass representing DNF package information."""
```

Comment:
I have used data classes before (and I prefer to use them!), but the reason why I have not used the dataclass decorator from [dataclasses](https://docs.python.org/3/library/dataclasses.html) in this case is because I am unsure what the RPM-based support story is for Juju and charmed operator. The EL7 and EL8 series of distributions only come with Python 3.6 out of the box, and dataclasses was not added to Python until version 3.8. This why I held off on using the dataclasses module. I had [similar issue](https://docs.python.org/3/library/dataclasses.html) in cleantest when I started adding support for more distributions.

If I can get charmed operators to work on a manually provisioned EL9 machine (they come with Python 3.9.14 out of the box), then I will look into using dataclasses instead. Until then I would prefer to not use the dataclasses module.

Reply:
We've just gone all in on Python 3.8+ as of `ops` 2.0 (see [release notes](https://github.com/canonical/operator/releases/tag/2.0.0)), as that's what is available on Focal. It'd be great if we can use EL9 if at all possible, so that we're on at least the same minimum Python version as `ops`. Plus, even if it takes a bit more figuring out, I think it's worth doing to get dataclasses and bunch of other more recent features.

Reply:
Yeah, `ops` going to Python 3.8 is a current problem that I am working through with the folks who need to use CentOS. I am going to be looking into getting an EL9 base to work. It might be rather hacky at first but we'll get there :smile: